test_util/provide_aws.py

```diff
@@ -0,0 +1,215 @@
+        Args:
+            key_name (str): key name to be registered in AWS
+        Returns:
+            string of contents of key
+        """
+        key = self.ec2_client.create_key_pair(KeyName=key_name)
+        return key['KeyMaterial']
+
+    def delete_key_pair(self, key_name):
+        self.ec2_client.delete_key_pair(KeyName=key_name)
+
+    @retrying.retry(stop_max_delay=1200*1000,
+                    wait_fixed=3000,
+                    retry_on_exception=lambda ex: False,
+                    retry_on_result=lambda res: res is False)
+    def wait_for_cluster_creation(self):
+        logger.info('Waiting for cluster to be created...')
+        self.stack.reload()
+        current_status = self.stack.stack_status
+        if current_status == 'CREATE_IN_PROGRESS':
+            return False
+        if current_status != 'CREATE_COMPLETE':
+            raise Exception("Expected stack {} to have status CREATE_COMPLETE,"
+                            "but found: {}".format(self.stack.stack_id,
+                                                   current_status))
+
+    @retrying.retry(stop_max_delay=1200*1000,
+                    wait_fixed=3000,
+                    retry_on_exception=lambda ex: False,
+                    retry_on_result=lambda res: res is False)
+    def wait_for_cluster_deletion(self):
+        logging.info('Waiting for cluster to be deleted...')
+        current_status = self.stack.stack_status
+        if current_status == 'DELETE_IN_PROGRESS':
+            return False
+        try:
+            self.stack.reload()
+        except botocore.exceptions.ClientError as ex:
+            resp = ex.response['Error']
+            if (
+                    resp['Error']['Code'] == 'ValidationError' and
+                    resp['Error']['Message'].endswith('does not exist')):
+                logging.info("Boto cannot find stack; deletion successful")
+                return True
+            else:
+                raise ex
+        if current_status != 'DELETE_COMPLETE':
+            raise Exception('Delete finished, but cluster has unexpected state: '+current_status)
+
+
+class CloudFormation(AwsProvision):
+    def create(self, name, template_url,
+               private_agents, public_agents,
+               admin_location, key_pair_name,
+               pre_1_8_cluster=False):
+        logging.info("Creating stack: {}".format(name))
+        parameters = [{
+                     'ParameterKey': 'KeyName',
+                     'ParameterValue': key_pair_name},
+                     {
+                     'ParameterKey': 'AdminLocation',
+                     'ParameterValue': admin_location},
+                     {
+                     'ParameterKey': 'PublicSlaveInstanceCount',
+                     'ParameterValue': str(public_agents)},
+                     {
+                     'ParameterKey': 'SlaveInstanceCount',
+                     'ParameterValue': str(private_agents)}]
+        if pre_1_8_cluster:
```

Comment:
this code I don't think will ever be used for Pre-1.8 clusters, can probably just drop this.